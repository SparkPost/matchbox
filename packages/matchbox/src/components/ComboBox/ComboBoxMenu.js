import React from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import { ActionList } from '../ActionList';
import styles from './ComboBoxMenu.module.scss';

function ComboBoxMenu(props) {
  const { items, menuRef, isOpen, maxHeight, ...rest } = props;

  const listClasses = classnames(
    styles.List,
    isOpen && styles.open
  );

  return (
    <div {...rest} ref={menuRef} className={listClasses}>
      <ActionList actions={items} maxHeight={maxHeight} />
    </div>
  );
}

ComboBoxMenu.propTypes = {
  /**
   * Array of items. see ActionList prop types.
   */
  items: PropTypes.array,
  /**
   * Controls menu visibility
   */
  isOpen: PropTypes.bool,
  /**
   * Maps to Downshift's refKey set in getMenuProps. refKey must be set to "menuRef"
   */
  menuRef: PropTypes.func
};

ComboBoxMenu.defaultProps = {
  items: []
};

export default ComboBoxMenu;
