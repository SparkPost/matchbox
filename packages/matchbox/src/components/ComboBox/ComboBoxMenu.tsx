import React from 'react';
import { MaxHeightProps } from 'styled-system';
import { ActionList, ActionListActionProps } from '../ActionList';
import { Box } from '../Box';
import PopoverContent from '../Popover/PopoverContent';

export type ComboBoxMenuProps = {
  /**
   * String displayed when item length is 0. Defaults to "No Results"
   */
  emptyMessage?: string;
  /**
   * Array of items. see ActionList prop types.
   */
  items?: ActionListActionProps[];
  /**
   * Controls menu visibility
   */
  isOpen?: boolean;
  /**
   * Maps to Downshift's refKey set in getMenuProps. refKey must be set to "menuRef"
   */
  menuRef?: React.MutableRefObject<HTMLDivElement>;
} & MaxHeightProps;

function ComboBoxMenu(props: ComboBoxMenuProps): JSX.Element {
  const {
    items = [],
    menuRef,
    isOpen,
    maxHeight = '20rem',
    emptyMessage = 'No Results',
    ...rest
  } = props;

  return (
    <Box ref={menuRef} {...rest} position="relative" zIndex="overlay">
      <PopoverContent open={isOpen} width="100%">
        {items.length > 0 ? (
          <ActionList maxHeight={maxHeight}>
            {items.map((action, i) => (
              <ActionList.Action key={i} {...action} />
            ))}
          </ActionList>
        ) : (
          <Box p="200" color="gray.700">
            {emptyMessage}
          </Box>
        )}
      </PopoverContent>
    </Box>
  );
}

export default ComboBoxMenu;
