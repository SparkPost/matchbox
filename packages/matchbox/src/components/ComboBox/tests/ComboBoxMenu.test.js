import React from 'react';
import ComboBoxMenu from '../ComboBoxMenu';
import { mount } from 'enzyme';

describe('ComboBoxMenu', () => {
  const subject = (props = {}) => mount(<ComboBoxMenu {...props}/>);
  const items = [
    { content: 'foo' },
    { content: <div>bar</div> }
  ];

  it('should render items correctly', () => {
    expect(subject({ items })).toMatchSnapshot();
  });

  it('should render correctly when open', () => {
    const menu = subject({ items, isOpen: true });
    expect(menu.find('.List')).toHaveAttributeValue('class', 'List open');
  });
});
