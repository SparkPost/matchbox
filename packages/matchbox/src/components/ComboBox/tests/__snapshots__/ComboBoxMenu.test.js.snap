// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ComboBoxMenu should render items correctly 1`] = `
<ComboBoxMenu
  items={
    Array [
      Object {
        "content": "foo",
      },
      Object {
        "content": <div>
          bar
        </div>,
      },
    ]
  }
>
  <div
    className="List"
  >
    <ActionList
      actions={
        Array [
          Object {
            "content": "foo",
          },
          Object {
            "content": <div>
              bar
            </div>,
          },
        ]
      }
      groupByKey="section"
    >
      <div
        className="ActionList"
        style={
          Object {
            "maxHeight": "none",
          }
        }
      >
        <Section
          key="0"
          section={
            Array [
              Object {
                "content": "foo",
              },
              Object {
                "content": <div>
                  bar
                </div>,
              },
            ]
          }
        >
          <div
            className="Section"
          >
            <UnstyledLink
              className="Action"
              key="0"
            >
              <a
                className="Action"
              >
                foo
              </a>
            </UnstyledLink>
            <UnstyledLink
              className="Action"
              key="1"
            >
              <a
                className="Action"
              >
                <div>
                  bar
                </div>
              </a>
            </UnstyledLink>
          </div>
        </Section>
      </div>
    </ActionList>
  </div>
</ComboBoxMenu>
`;
